README.md
## Waht is dis
#### why is there haskell and html in a file??
I needed a site

#### but why html
my browser still only support html :( (le sad)

#### but why haskell
I rolled my dnd dice and got an 8 (eight). This is the letter H so I googled it and Haskell was the first result (it wasnt i just need an excuse, its cool ok)

#### does this site reflect your or your employers political views
but why?

## Where is?
# http://slavecolony.hopto.org




generate-hs/CHANGELOG.md
# Revision history for generate-hs

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


generate-hs/app/Main.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Web.Scotty
import Data.Aeson (FromJSON, ToJSON, object, (.=), decode, encode)
import Data.Time.Clock ( getCurrentTime )
import Control.Monad.IO.Class (liftIO)
import Data.Maybe (Maybe)

-- owned
import Repository
import HtmlGen ( patchIndexWith )
import Structure ( Post(..))

instance ToJSON Post
instance FromJSON Post

main :: IO ()
main = do
  repository <- initializeRepository

  scotty 3000 $
    get "/" helloHaskellHandler >>
    get "/site-logo.png" (fileHandler "site-logo.png") >>
    get "/pylon.css" (fileHandler "pylon.css") >>
    get "/:id" (postIdParamHandler repository) >>
    post "/newpost" (newPostHandler repository) >>
    get "/all" (getAllPostsHandler repository) >>
    post "/regenerate" (regenerateHandler repository) >>
    delete "/post" (deletePostHandler repository)

  closeRepository repository


-- Route Handlers
helloHaskellHandler :: ActionM ()
helloHaskellHandler = file "index-u.html"

fileHandler :: FilePath -> ActionM ()
fileHandler = file

postIdParamHandler :: Repository -> ActionM ()
postIdParamHandler repo = do
  postId <- captureParam "id"
  post <- liftIO $ getPostById repo postId
  json post

newPostHandler :: Repository -> ActionM ()
newPostHandler repo = do
  newPost <- jsonData :: ActionM Post
  timestamp <- liftIO getCurrentTime
  let postWithTimestamp = newPost { timestamp = Just timestamp }
  postId <- liftIO $ insertPost repo postWithTimestamp
  json $ object ["id" .= postId]

getAllPostsHandler :: Repository -> ActionM ()
getAllPostsHandler repo = do
  posts <- liftIO $ getAllPosts repo
  json posts

regenerateHandler :: Repository -> ActionM ()
regenerateHandler repo = do
  posts <- liftIO $ getAllPublicPosts repo
  liftIO $ patchIndexWith posts
  file "index-u.html"

deletePostHandler :: Repository -> ActionM ()
deletePostHandler repo = do
  id <- queryParam "id"
  liftIO $ setToPrivate repo id


generate-hs/app/Repository.hs
{-# LANGUAGE OverloadedStrings #-}

module Repository (
  Repository,
  initializeRepository,
  closeRepository,
  getPostById,
  insertPost,
  getAllPosts,
  getAllPublicPosts,
  setToPrivate
) where

import Structure (Post(..))
import Database.SQLite.Simple
import Data.Time.Clock
import qualified Data.Text as T
import GHC.Generics
import Data.Aeson.Key (toString)
import Data.Text (Text)

instance FromRow Post where
  fromRow = Post <$> field <*> field <*> field <*> field <*> field

instance ToRow Post where
  toRow (Post id title content timestamp link) = toRow (title, content, timestamp, link, 0 :: Int)

-- Define a Repository data type to encapsulate database operations
newtype Repository = Repository Connection

buildQuery :: Text -> Text -> Query
buildQuery = (Query .) . (<>)

selectPosts :: Text -> Query
selectPosts = buildQuery "SELECT id, title, content, timestamp, link FROM posts "

setPostPrivate :: Text -> Query
setPostPrivate = buildQuery "UPDATE posts SET private"

initializeRepository :: IO Repository
initializeRepository = do
  conn <- open "posts.db"
  execute_ conn "CREATE TABLE IF NOT EXISTS posts (id INTEGER PRIMARY KEY, title TEXT, content TEXT, timestamp TIMESTAMP, link TEXT, private INTEGER)"
  return $ Repository conn

closeRepository :: Repository -> IO ()
closeRepository (Repository conn) = close conn

getPostById :: Repository -> Int -> IO (Maybe Post)
getPostById (Repository conn) postId = do
  result <- query conn (selectPosts "WHERE id = ?") (Only postId)
  return $ case result of
    [post] -> Just post
    _      -> Nothing

insertPost :: Repository -> Post -> IO Int
insertPost (Repository conn) post = do
  execute conn "INSERT INTO posts (title, content, timestamp, link, private) VALUES (?, ?, ?, ?, ?)" post
  fromIntegral <$> lastInsertRowId conn

getAllPosts :: Repository -> IO [Post]
getAllPosts (Repository conn) = query_ conn $ selectPosts ""

getAllPublicPosts :: Repository -> IO [Post]
getAllPublicPosts (Repository conn) = query_ conn $ selectPosts "WHERE private = 0" 

setToPrivate :: Repository -> Int -> IO ()
setToPrivate (Repository conn) id = execute conn (setPostPrivate " = 1 WHERE id = ?") (Only id)

setToPublic :: Repository -> Int -> IO ()
setToPublic (Repository conn) id = execute conn (setPostPrivate "= 0 WHERE id = ?") (Only id)


generate-hs/app/Html-generation.hs
module Main where

import Text.Blaze.Html5 as H
import Text.Blaze.Html5.Attributes as A
import Text.Blaze.Html.Renderer.String (renderHtml)
import Data.Time.Clock
import Data.Function
import Data.Time (formatTime, defaultTimeLocale)

(|>) = (&)

htmlclass :: String -> Attribute
htmlclass str = A.class_ (toValue str)

data Post = Post {
  postId :: Int,
  postTitle :: String,
  postContent :: String,
  postTimestamp :: Maybe UTCTime
} deriving (Show)

formatUTCTime :: Maybe UTCTime -> Maybe String
formatUTCTime (Just time) = Just $ formatTime defaultTimeLocale "%d-%m-%Y" time
formatUTCTime _ = Nothing

-- Function to generate HTML for a single post
-- Convert a Post to HTML
postToHtml :: Post -> Html
postToHtml post =
  H.div ! htmlclass "card" $
    H.div ! htmlclass "info" $
      do
        H.div ! htmlclass "card-title" $
          do
            H.h3 ! htmlclass "title" $ toHtml $ postTitle post
            H.p ! htmlclass "timestamp" $ maybe mempty toHtml (formatUTCTime $ postTimestamp post)
        H.p ! htmlclass "content" $ toHtml $ postContent post

-- Function to generate HTML for a list of posts and wrap them in a div with id "posts"
postsToHtml :: [Post] -> Html
postsToHtml = foldMap postToHtml

-- Example usage
main :: IO ()
main =
  [     Post 1 "Title 1" "Content 1" (Just $ read "2023-01-01 12:00:00 UTC"),
        Post 2 "Title 2" "Content 2" Nothing,
        Post 3 "Title 3" "Content 3" (Just $ read "2023-01-03 15:30:00 UTC")
  ] |> postsToHtml 
  |> renderHtml 
  |> putStrLn

generate-hs/app/HtmlGen.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE OverloadedStrings #-}


module HtmlGen(
  patchIndexWith
) where

import Text.Blaze.Html5 as H hiding (title)
import Text.Blaze.Html5.Attributes as A hiding (title, content)
import Text.Blaze.Html.Renderer.String (renderHtml)
import Data.Time.Clock
import Data.Function
import Data.Time (formatTime, defaultTimeLocale)
import qualified Data.Text as T
import qualified Data.Text.Lazy as TL
import Text.Markdown (markdown, def)

import Structure (Post(..))

(|>) :: a -> (a -> b) -> b
(|>) = (&)

htmlclass :: String -> Attribute
htmlclass str = A.class_ (toValue str)

formatUTCTime :: Maybe UTCTime -> Maybe String
formatUTCTime (Just time) = Just $ formatTime defaultTimeLocale "%d-%m-%Y" time
formatUTCTime _ = Nothing

markdownToHtml :: TL.Text -> Html
markdownToHtml = preEscapedToHtml . markdown def

-- Function to generate HTML for a single post
-- Convert a Post to HTML
postToHtml :: Post -> Html
postToHtml post =
  H.div ! htmlclass "card" $
    H.div ! htmlclass "info" $
      do
        H.div ! htmlclass "card-title" $
          do
            H.h3 ! htmlclass "title" $ toHtml $ title post
            H.p ! htmlclass "timestamp" $ maybe mempty toHtml (formatUTCTime $ timestamp post)
        toHtml (markdownToHtml $ TL.pack $ content post) ! htmlclass "content"
        H.a ! htmlclass "link" ! A.style "width: 3.5rem;" ! A.href "http://example.com" $ "Red pill"

-- Function to generate HTML for a list of posts and wrap them in a div with id "posts"
postsToHtml :: [Post] -> Html
postsToHtml = foldMap postToHtml

patchIndexWith :: [Post] -> IO ()
patchIndexWith posts = do
  index <- readFile "index.html"
  let insertionPortion = T.replace "<!-- placeholder -->" ("<div class=\"timeline\"><div class=\"outer\">" <> T.pack (renderHtml (postsToHtml posts) <> "</div></div>")) (T.pack index)
  (putStrLn . T.unpack) insertionPortion
  writeFile "index-u.html" (T.unpack insertionPortion)

-- Example usage
main :: IO ()
main = {-
  ([     Post 1 "Title 1" "Content 1" (Just $ read "2023-01-01 12:00:00 UTC"),
        Post 2 "Title 2" "Content 2" Nothing,
        Post 3 "Title 3" "Content 3" (Just $ read "2023-01-03 15:30:00 UTC")
  ] |> postsToHtml
  |> renderHtml
  |> putStrLn) >> -} ([     Post 1 "Title 1" "Content 1 generated  Long style content who even has to say that much like wtf dont bother me with this waste of sapce" (Just $ read "2023-01-01 12:00:00 UTC") Nothing,
        Post 2 "Title 2" "Content 2 generated" Nothing Nothing,
        Post 3 "Title 3" "Content 3 generated" (Just $ read "2023-01-03 15:30:00 UTC") Nothing
  ] |> patchIndexWith
  )
